### PR TITLE
Add insecure and plain-http flag when pull and push

### PR DESCRIPTION
**Pull request by pytimer**
This is the first commit about pull/push to an insecure registry.  

This PR implement pull or push to an insecure registry using the CNAME, but using the ip address is wrong. So this PR setting WIP. 

If someone know why, please tell me, thanks.

TODO:
- [x] pull/push to an insecure registry using the ip address, such as `192.168.100.20:8443`.
- [x] add `--plain-http` to pull or push a HTTP registry.

Related issue:
- https://github.com/helm/helm/issues/6324

### CODE REVIEW

**Comment by jdolitsky** [comment]
@pytimer this looks good so far. If we could add some section to your part in the readme like "please use --insecure on push and pull"

**Comment by pytimer** [comment]
@jdolitsky This insecure flag is invalid for using ip address to access an insecure registry, do you know why?

I have no idea about it, so I think it maybe can not merge if this question not solved.

**Comment by pytimer** [comment]
@jdolitsky I finish this feature, hope you can review it.

I also update the README.md about this feature usage, my english not good, so if not described clearly,  please point out.

File: cmd/oras/push.go [hunk 1]
```diff
@@ -31,10 +31,12 @@ type pushOptions struct {
...
-	debug    bool
-	configs  []string
-	username string
-	password string
+	debug     bool
+	configs   []string
+	username  string
+	password  string
+	insecure  bool
+	PlainHTTP bool
```
Comment (jdolitsky) [inline cmd/oras/push.go]: Can we follow the pattern and keep this private? `plainHTTP`?

File: README.md [hunk 2]
```diff
@@ -173,6 +173,56 @@ To remove the entry from the credentials file, use `oras logout`:
...
+    -v $(pwd)/auth.htpasswd:/etc/docker/registry/auth.htpasswd \
+    -e REGISTRY_AUTH="{htpasswd: {realm: localhost, path: /etc/docker/registry/auth.htpasswd}}" \
+    -e REGISTRY_HTTP_ADDR=0.0.0.0:5000 \
+    registry
+```
+
+In a new window, login with `oras` using the ip address not localhost:
+
+```
+oras login -u myuser -p mypass --insecure <registry-ip>:5000
```
Comment (jdolitsky) [inline README.md]: This is misleading - if the registry is not protected by https, why do we need insecure flag?

File: README.md [hunk 3]
```diff
@@ -173,6 +173,56 @@ To remove the entry from the credentials file, use `oras logout`:
...
+```
+
+You will notice a new entry for `<registry-ip>:5000` appear in `~/.docker/config.json`.
+
+Then you can pull files from the registry or push files to the registry.
+
+To push single file to this registry:
+
+```
+oras push --insecure <registry-ip>:5000/library/hello:latest hi.txt --plain-http
```
Comment (jdolitsky) [inline README.md]: same here - I would expect to only ever use one or the other (`--insecure` or `--plain-http`)

File: cmd/oras/push.go (hunk 1 above)
Comment (pytimer) [reply cmd/oras/push.go]: Fix it.

**Comment by pytimer** [comment]
@jdolitsky Sorry, my example not descirbed clearly, i update the README.md, you can see again.

The `--plain-http` used when using with the plain HTTP registry, `--insecure` used when using with the insecure HTTPS registry.

File: README.md (hunk 2 above)
Comment (pytimer) [reply README.md]: The registry go modules support `InsecureRegistries` option to first using HTTPS, if HTTPS is not available, fall back to HTTP. so the plain HTTP registry or insecure HTTPS registry, always using the `--insecure` flag to login.

I have considered pull/push also using the `--insecure` flag, but i not found it support this option, so i add `--plain-http` flag when pull/push now.

**Comment by zygon4** [comment]
@pytimer is the entire commit for this PR held back due to this conflict? Is anyone looking at finishing the merge?

Thanks!

**Comment by pytimer** [comment]
@zygon4 I waiting review completed. 

@jdolitsky @zygon4 I solve the conflict files and move documents to implementors.md, please review them. Thanks.

**Comment by zygon4** [comment]
> 
> 
> @zygon4 I waiting review completed.
> 
> @jdolitsky @zygon4 I solve the conflict files and move documents to implementors.md, please review them. Thanks.

Thanks so much! I don't think I'm qualified to review, however. @jdolitsky ?

Thanks again!

**Comment by jdolitsky** [comment]
thanks again @pytimer !

**Pull request merged**